app/processing/pre/io_parser.py

```diff
@@ -0,0 +1,176 @@
+from io import UnsupportedOperation
+
+from openqasm3.ast import (
+    AliasStatement,
+    Annotation,
+    Concatenation,
+    Expression,
+    Identifier,
+    IndexExpression,
+    QASMNode,
+    QubitDeclaration,
+)
+
+from app.openqasm3.visitor import LeqoTransformer
+from app.processing.graph import (
+    IOInfo,
+    SingleInputInfo,
+    SingleIOInfo,
+    SingleOutputInfo,
+)
+from app.processing.utils import expr_to_int, parse_io_annotation, parse_qasm_index
+
+
+class IOParse(LeqoTransformer[None]):
+    """Parse input/output qubits of a single qasm-snippet.
+
+    Do it the following way:
+    - give every declared qubit (not qubit-reg) and id, based on declaration order
+    - create map that points from declared identifiers to ids, this also parses aliases
+    - store annotation info based on the id of the qubit
+    """
+
+    qubit_id: int
+    input_counter: int
+    output_counter: int
+    io: IOInfo
+
+    def __init__(self, io: IOInfo) -> None:
+        """Construct the LeqoTransformer.
+
+        :param io: The IOInfo to be modified in place.
+        """
+        super().__init__()
+        self.qubit_id = 0
+        self.input_counter = 0
+        self.output_counter = 0
+        self.io = io
+
+    def visit_QubitDeclaration(self, node: QubitDeclaration) -> QASMNode:
+        """Parse qubit-declarations an there corresponding input annotations."""
+        name = node.qubit.name
+        size = expr_to_int(node.size) if node.size is not None else 1
+        input_id: int | None = None
+        dirty = False
+        for annotation in node.annotations:
+            match annotation.keyword:
+                case "leqo.input":
+                    if input_id is not None or dirty:
+                        msg = f"Unsuported: two input/dirty annotations over {name}"
```

Comment:
It would be nice to have clearer error messages (Not multiple possible cases in one message)